craft_parts/plugins/maven_plugin.py

```diff
@@ -0,0 +1,191 @@
+
+
+class MavenPluginEnvironmentValidator(validator.PluginEnvironmentValidator):
+    """Check the execution environment for the maven plugin.
+
+    :param part_name: The part whose build environment is being validated.
+    :param env: A string containing the build step environment setup.
+    """
+
+    def validate_environment(self, *, part_dependencies: Optional[List[str]] = None):
+        """Ensure the environment contains dependencies needed by the plugin.
+
+        :param part_dependencies: A list of the parts this part depends on.
+
+        :raises PluginEnvironmentValidationError: If go is invalid
+        and there are no parts named go.
+        """
+        version = self.validate_dependency(
+            dependency="mvn",
+            plugin_name="maven",
+            part_dependencies=part_dependencies,
+        )
+        if not re.match(r"\S*Apache Maven ", version) and (
+            part_dependencies is None or "maven-deps" not in part_dependencies
+        ):
+            raise errors.PluginEnvironmentValidationError(
+                part_name=self._part_name,
+                reason=f"invalid maven version {version!r}",
+            )
+
+
+class MavenPlugin(JavaPlugin):
+    """A plugin to build parts that use Maven.
+
+    The Maven build system is commonly used to build Java projects. This
+    plugin requires a pom.xml in the root of the source tree.
+
+    This plugin uses the common plugin keywords as well as those for "sources".
+    For more information check the 'plugins' topic for the former and the
+    'sources' topic for the latter.
+
+    Additionally, this plugin uses the following plugin-specific keywords:
+
+        - maven-parameters:
+          (list of strings)
+          Flags to pass to the build using the maven semantics for parameters.
+    """
+
+    properties_class = MavenPluginProperties
+    validator_class = MavenPluginEnvironmentValidator
+
+    def get_build_snaps(self) -> Set[str]:
+        """Return a set of required snaps to install in the build environment."""
+        return set()
+
+    def get_build_packages(self) -> Set[str]:
+        """Return a set of required packages to install in the build environment."""
+        return set()
+
+    def get_build_environment(self) -> Dict[str, str]:
+        """Return a dictionary with the environment to use in the build step."""
+        return {}
+
+    def get_build_commands(self) -> List[str]:
+        """Return a list of commands to run during the build step."""
+        options = cast(MavenPluginProperties, self._options)
+
+        mvn_cmd = ["mvn", "package"]
+        if self._use_proxy():
+            settings_path = self._part_info.part_build_dir / ".parts/m2/settings.xml"
+            _create_settings(settings_path)
+            mvn_cmd += ["-s", str(settings_path)]
+
+        return [
+            " ".join(mvn_cmd + options.maven_parameters)
+        ] + self._get_java_post_build_commands()
+
+    def _use_proxy(self):
```

Comment:
```suggestion
    def _use_proxy(self) -> bool:
```

Maybe we should add `disallow_untyped_defs` on new files? Or just bite the bullet and annotate the whole codebase

Reply:
I think we can start some background work to move in that direction instead of doing it all at once. How do you enforce it only for new files?